### PR TITLE
Added Support for Lazy Quantifier [databricks]

### LINKED ISSUE #8806: [FEA] Support lazy quantifier and specified group index in regexp_extract function
It would be nice if we could support lazy quantifier and specified group index in regexp_extract function.
For example:
```
create a hive table and insert some rows:
hive> show create table datavalid3;
OK
CREATE EXTERNAL TABLE `datavalid3`(
  `col1` int,
  `col2` bigint,
  `col3` tinyint,
  `col4` string,
  `col5` date,
  `col6` map<string,string>,
  `col7` map<string,array<string>>,
  `col8` decimal(5,3))
PARTITIONED BY (
  `col9` string)
ROW FORMAT SERDE
  'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
STORED AS INPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
OUTPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat'

run below SQL:
spark.sql("select regexp_extract(col4,'\"xst\":\"(.*?)\"') from datavalid3").show
```
fallback info:
```
!Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
    @Expression <Alias> regexp_extract(col4#3, "xst":"(.*?)", 1) AS regexp_extract(col4, "xst":"(.*?)", 1)#21 could run on GPU
      !Expression <RegExpExtract> regexp_extract(col4#3, "xst":"(.*?)", 1) cannot run on GPU because regex group count is 0, but the specified group index is 1; Lazy quantifier *? not supported near index 10
```

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds the lazy quantifier to the RegexParser 

**Changes Made:** 
* RegexParser returns the needed transpiled AST needed
* Test was added to regexp_test.py 

**Tests:** 
* Integration test added to regexp_test.py

Integration tests were run locally

fixes #8806

### CODE REVIEW

**Comment by razajafri** [comment]
I will be adding more fuzzer tests

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
Databricks cluster failed to start.

[2024-01-17T21:16:34.414Z] subprocess.CalledProcessError: Command 'ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -p 2200 -i **** ubuntu@4.155.42.241 'SPARKSRCTGZ=/home/ubuntu/spark-rapids-ci.tgz BASE_SPARK_VERSION=3.2.1 BASE_SPARK_VERSION_TO_INSTALL_DATABRICKS_JARS=3.2.1 MVN_OPT= EXTRA_ENVS=         bash /home/ubuntu/build.sh  2>&1 | tee buildout; if [ `echo ${PIPESTATUS[0]}` -ne 0 ]; then false; else true; fi'' returned non-zero exit status 1.

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 1]
```diff
@@ -1516,9 +1516,10 @@ class CudfRegexTranspiler(mode: RegexMode) {
         case (RegexRepetition(_, SimpleQuantifier('*')), SimpleQuantifier('+')) =>
           throw new RegexUnsupportedException("Possessive quantifier *+ not supported",
             quantifier.position)
-        case (RegexRepetition(_, SimpleQuantifier('*')), SimpleQuantifier('?')) =>
-          throw new RegexUnsupportedException("Lazy quantifier *? not supported",
-            quantifier.position)
+        case (RegexRepetition(_, SimpleQuantifier('?')) |
+              RegexRepetition(_, SimpleQuantifier('*')) |
+              RegexRepetition(_, SimpleQuantifier('+')), SimpleQuantifier('?')) =>
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1521]: nit: the pattern matching can be more concise here

```suggestion
        case (RegexRepetition(_, SimpleQuantifier('?' | '*' | '+')), SimpleQuantifier('?')) =>
```

**Review by andygrove — APPROVED** [review]
LGTM. I ran this locally and manually verified that the results from the integration test looked correct. I also tried some different datagen patterns to see if I could break the test but did not find anything.

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
If Databricks keeps failing I will just run the test manually and skip Databricks from the premerge

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by sameerz** [comment]
build

**Comment by NVnavkumar** [comment]
Looking at recent blossom failure: 

```
2024-01-22T01:00:54.1200895Z [2024-01-22T00:55:30.814Z] [31m- AST fuzz test - regexp_find *** FAILED ***[0m
2024-01-22T01:00:54.1201603Z [2024-01-22T00:55:30.815Z] [31m  cuDF failed to compile pattern: (2|\u2029??)+? (RegularExpressionTranspilerSuite.scala:954)
```

It looks like cuDF doesn't support the `??` or `?` inside of the choice( `|`). Will need to handle that case explicitly and fall back to CPU. The `+?` after the group is fine.

**Comment by razajafri** [comment]
> Looking at recent blossom failure:
> 
> ```
> 2024-01-22T01:00:54.1200895Z [2024-01-22T00:55:30.814Z] �[31m- AST fuzz test - regexp_find *** FAILED ***�[0m
> 2024-01-22T01:00:54.1201603Z [2024-01-22T00:55:30.815Z] �[31m  cuDF failed to compile pattern: (2|\u2029??)+? (RegularExpressionTranspilerSuite.scala:954)
> ```
> 
> It looks like cuDF doesn't support the `??` or `?` inside of the choice( `|`). Will need to handle that case explicitly and fall back to CPU. The `+?` after the group is fine.

Thanks for looking into this. I am struggling to see why the pattern is being generated

**Comment by NVnavkumar** [comment]
> > Looking at recent blossom failure:
> > ```
> > 2024-01-22T01:00:54.1200895Z [2024-01-22T00:55:30.814Z] �[31m- AST fuzz test - regexp_find *** FAILED ***�[0m
> > 2024-01-22T01:00:54.1201603Z [2024-01-22T00:55:30.815Z] �[31m  cuDF failed to compile pattern: (2|\u2029??)+? (RegularExpressionTranspilerSuite.scala:954)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It looks like cuDF doesn't support the `??` or `?` inside of the choice( `|`). Will need to handle that case explicitly and fall back to CPU. The `+?` after the group is fine.
> 
> Thanks for looking into this. I am struggling to see why the pattern is being generated

It's because it's actually a valid pattern in Java (and Python fwiw). CUDF is the only thing throwing an invalid pattern here. Since it's valid in Java, and we have not detected to fallback to CPU, that's why you are getting the test error here.

**Comment by andygrove** [comment]
I was able to get past the failing test with this diff. Note that this is only checking the right side of the choice and we would need to look at the left as well.

```
$ git diff
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
index 2380eb623..7666f1ab9 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
@@ -1565,6 +1565,14 @@ class CudfRegexTranspiler(mode: RegexMode) {
             r.position)
         }
 
+        rr match {
+          case RegexSequence(ListBuffer(RegexRepetition(
+              RegexRepetition(_,SimpleQuantifier('?')),SimpleQuantifier('?')))) =>
+            throw new RegexUnsupportedException(
+              "cuDF does not support lazy quantifier inside choice", r.position)
+          case other =>
+        }
+
         RegexChoice(ll, rr)
```

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
Thanks @andygrove

**Comment by razajafri** [comment]
There are more test failures, I am looking into them

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 2]
```diff
@@ -1566,6 +1565,17 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
         }
 
+        (ll, rr) match {
+          // ll = lazyQuantifier inside a choice
+          case (RegexSequence(ListBuffer(RegexRepetition(
+          RegexRepetition(_, SimpleQuantifier('?')), SimpleQuantifier('?')))), _) |
+               // rr = lazyQuantifier inside a choice
+               (_, RegexSequence(ListBuffer(RegexRepetition(
+               RegexRepetition(_, SimpleQuantifier('?')), SimpleQuantifier('?'))))) =>
+            throw new RegexUnsupportedException(
```
Comment (sameerz) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1575]: If we are creating exceptions, we need to document these in https://github.com/NVIDIA/spark-rapids/blob/branch-24.02/docs/compatibility.md#regular-expressions

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -1566,6 +1565,20 @@ class CudfRegexTranspiler(mode: RegexMode) {
             r.position)
         }
 
+        (ll, rr) match {
+          // ll = lazyQuantifier inside a choice
+          case (RegexSequence(ListBuffer(RegexRepetition(
+          RegexRepetition(_, SimpleQuantifier('?')), SimpleQuantifier('?')))), _) |
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1571]: nit: indent seems off here? I would expect this second line to be indented more that the `case` in the prev line. Same comment for other uses of this pattern in this PR.

**Review by andygrove — APPROVED** [review]
LGTM. Lett a comment about formatting.

**Pull request merged**